kubernetes-deployment/src/main.tf

```diff
@@ -0,0 +1,59 @@
+# TODO: chart name comes from var.md_metadata.name_prefix
+# TODO: make sure the 'host' field is actually just the subdomain, k8s will be handling the zone. Make this clear in the description
+# TODO: I dont see any external-dns in the chart...
+
+module "application" {
+  source = "github.com/massdriver-cloud/terraform-modules//massdriver-application"
+  name   = var.md_metadata.name_prefix
+
+  # TODO: this will need to be finagled per cloud type to get the right k8s bits.
+  # This could be moved into the module w/ a simpler api here like
+  # service = "ec2" | "lambda" | "k8s:the-data-in-some-consisten-way-that-the-module-can-parse"
+  identity = {
+    assume_role_policy = <<EOF
```

Comment:
@WillBeebe @chrisghill It would be great to see an example of what the values are for GCP and Azure for k8s. Can you put them here as a comment? I want to get this abstracted down to just a `service` field and let the module handle all of this jazz... 

AWS for example...

```hcl
service = "lambda"
service = "ec2"
service = "k8s:$OIDC:$NAMESPACE:$NAME"
```

Each of these would produce the correct assume role policy, lambda/ec2 would use canned. The k8s would produce:

```hcl
  assume_role_policy = jsonencode({
    Version = "2012-10-17"
    Statement = [{
      "Sid"    = "EksCertManager"
      "Effect" = "Allow",
      "Principal" = {
        "Federated" = "arn:aws:iam::${data.aws_caller_identity.current.account_id}:oidc-provider/${local.eks_oidc_short}"
      }
      "Action" = "sts:AssumeRoleWithWebIdentity",
      "Condition" = {
        "StringEquals" = {
          "${local.eks_oidc_short}:sub" = "system:serviceaccount:${var.namespace}:${var.name}"
        }
      }
    }]
  })
```

@chrisghill we may need to include `account_id` (or a good generic name) as an output on `mdxc_cloud` .

Reply:
> AWS for example...
```
service = "lambda"
service = "ec2"
service = "k8s:$OIDC:$NAMESPACE:$NAME"
```

We can also just have a `kubernetes` block that is `required` if the `service` = kubernetes` or something like that instead of having some sort of delimited string.

AWS needs:
* AWS Account ID
* OIDC Issuer URL - the URL *might* be able to be calculated, aside from an ID field unrelated to the cluster name that is a random string (a bit like Route53 Hosted Zone ID). Its probably easier to just pass the whole URL since it has the ID in it.
* Kubernetes Namespace
* Kubernetes Service Account Name

Reply:
LMK what azure and GCP looks like

Reply:
Azure doesn't support workload identity yet, so its static credentials (comes out in ENVS).
GCP is this I think: https://github.com/massdriver-cloud/terraform-modules/blob/main/k8s-cert-manager-gcp/iam.tf#L55

* Project ID
* Namespace
* Service Account

@WillBeebe will need to verify if he needs anything else.

Reply:
What is this 1763? They haven’t released it yet?

Reply:
Its in public preview. I think Michael asked around and its expected to go GA late this year.

We attempted to add it to AKS cluster but there were multiple issues trying to do it w/ terraform. Required opting into preview features in the API (hard to do with TF) and a installing CRDs and admission controllers in the cluster.  We kinda got it working, but the Azure SDKs didn't support the credential chain yet, so services (like cert-manager and external-dns) didn't fall back to the dynamic credentials anyway and required static creds.

Reply:
V v cool

Reply:
I should be clear that the Azure SDKs probably *do* support the credential chains, but services need to be updated to use the SDKs to properly execute the credential chain. Some guy made tickets on both `cert-manager` and `external-dns` to do that:

https://github.com/cert-manager/cert-manager/issues/5085
https://github.com/kubernetes-sigs/external-dns/issues/2724

Reply:
@chrisghill you nailed it. 

- Namespace
- Project
- GCP service account email (have it already)


In the Terraform we built the k8s service account name like this: (you linked the last line, adding in-line to reference)

```terraform
resource "google_service_account_iam_member" "main" {
  service_account_id = google_service_account.application.name
  role               = "roles/iam.workloadIdentityUser"
  member             = "serviceAccount:${var.project_id}.svc.id.goog[${var.namespace}/${var.md_metadata.name_prefix}]"
}
```

We also need this passed to helm to annotate the service account. When I was playing with the mdxc provider I made this conditional on the cloud being `gcp`.

```terraform
locals {
  helm_additional_values_gcp = mdxc_application_identity.main.cloud == "gcp" ? {
     serviceAccount = {
       annotations = {
         "iam.gke.io/gcp-service-account" = mdxc_application_identity.main.ID
       }
     }
   } : {}
}
```